File: frontend/src/types/transformation.types.ts
export enum TransformationType {
  FILTER = 'filter',
  CLEAN = 'clean',
  AGGREGATE = 'aggregate',
  JOIN = 'join',
  SPLIT_COLUMN = 'split_column',
  MERGE_COLUMN = 'merge_column',
  CUSTOM_SQL = 'custom_sql',
  CUSTOM_PYTHON = 'custom_python',
  TYPE_CONVERSION = 'type_conversion',
  RENAME = 'rename',
  DROP = 'drop',
  FILL_NULL = 'fill_null',
}

export enum FilterOperator {
  EQUALS = '=',
  NOT_EQUALS = '!=',
  GREATER_THAN = '>',
  LESS_THAN = '<',
  GREATER_EQUAL = '>=',
  LESS_EQUAL = '<=',
  IN = 'in',
  NOT_IN = 'not_in',
  CONTAINS = 'contains',
  NOT_CONTAINS = 'not_contains',
  STARTS_WITH = 'starts_with',
  ENDS_WITH = 'ends_with',
  IS_NULL = 'is_null',
  NOT_NULL = 'not_null',
}

export enum AggregateFunction {
  SUM = 'sum',
  COUNT = 'count',
  COUNT_DISTINCT = 'count_distinct',
  AVG = 'avg',
  MIN = 'min',
  MAX = 'max',
  MEDIAN = 'median',
  STD = 'std',
  VAR = 'var',
}

export enum JoinType {
  INNER = 'inner',
  LEFT = 'left',
  RIGHT = 'right',
  FULL = 'full',
  CROSS = 'cross',
}

export interface FilterRule {
  column: string;
  operator: FilterOperator;
  value?: any;
  case_sensitive?: boolean;
}

export interface CleaningRule {
  column: string;
  rule_type: string;
  parameters?: Record<string, any>;
}

export interface AggregationConfig {
  group_by: string[];
  aggregations: Array<{
    column: string;
    function: string;
    alias?: string;
  }>;
  having?: FilterRule[];
}

export interface JoinConfig {
  left_source: {
    datasource_id?: number;
    table_name?: string;
    query?: string;
  };
  right_source: {
    datasource_id: number;
    table_name?: string;
    query?: string;
  };
  join_type: JoinType;
  join_conditions: Array<{
    left: string;
    right: string;
  }>;
}

export interface ColumnSplitConfig {
  column: string;
  delimiter?: string;
  pattern?: string;
  new_columns: string[];
  keep_original?: boolean;
}

export interface ColumnMergeConfig {
  columns: string[];
  separator?: string;
  new_column: string;
  drop_original?: boolean;
}

export interface TypeConversionConfig {
  column: string;
  target_type: string;
  format?: string;
  default_value?: any;
}

export interface TransformationStep {
  id?: string;
  name: string;
  type: TransformationType;
  config: Record<string, any>;
  description?: string;
}

export interface TransformationPipeline {
  id?: string;
  name: string;
  description?: string;
  source_config: {
    datasource_id: number;
    table_name?: string;
    query?: string;
  };
  steps: TransformationStep[];
  output_config?: {
    type: 'table' | 'export';
    datasource_id?: number;
    table_name?: string;
    format?: 'csv' | 'excel';
    filename?: string;
    if_exists?: 'replace' | 'append' | 'upsert' | 'merge' | 'fail';
    primary_key_columns?: string[];
  };
  created_at?: string;
  updated_at?: string;
}

export interface TransformationPreviewRequest {
  source_config: {
    datasource_id: number;
    table_name?: string;
    query?: string;
  };
  steps: TransformationStep[];
  preview_rows?: number;
}

export interface TransformationExecuteRequest {
  pipeline_id?: string;
  source_config?: {
    datasource_id: number;
    table_name?: string;
    query?: string;
  };
  steps?: TransformationStep[];
  output_config: {
    type: 'table' | 'export';
    datasource_id?: number;
    table_name?: string;
    format?: 'csv' | 'excel';
    filename?: string;
    if_exists?: 'replace' | 'append' | 'upsert' | 'merge' | 'fail';
    primary_key_columns?: string[];
  };
  execute_async?: boolean;
}

export interface TransformationPreviewResponse {
  status: string;
  original_shape: {
    rows: number;
    columns: number;
  };
  transformed_shape: {
    rows: number;
    columns: number;
  };
  columns: string[];
  data_types: Record<string, string>;
  preview: Record<string, any>[];
  transformations_applied: number;
}

export interface TransformationExecuteResponse {
  status: string;
  message: string;
  rows_processed: number;
  columns: string[];
  output_location: {
    type: string;
    datasource_id?: number;
    table_name?: string;
    format?: string;
    path?: string;
  };
}

File: backend/app/models/transformation_types.py
from typing import Any, List, Dict, Optional
from pydantic import BaseModel, Field
from enum import Enum


class TransformationType(str, Enum):
    FILTER = "filter"
    CLEAN = "clean"
    AGGREGATE = "aggregate"
    JOIN = "join"
    SPLIT_COLUMN = "split_column"
    MERGE_COLUMN = "merge_column"
    CUSTOM_SQL = "custom_sql"
    CUSTOM_PYTHON = "custom_python"
    TYPE_CONVERSION = "type_conversion"
    RENAME = "rename"
    DROP = "drop"
    FILL_NULL = "fill_null"


class FilterOperator(str, Enum):
    EQUALS = "="
    NOT_EQUALS = "!="
    GREATER_THAN = ">"
    LESS_THAN = "<"
    GREATER_EQUAL = ">="
    LESS_EQUAL = "<="
    CONTAINS = "contains"
    NOT_CONTAINS = "not_contains"
    STARTS_WITH = "starts_with"
    ENDS_WITH = "ends_with"
    IN = "in"
    NOT_IN = "not_in"
    IS_NULL = "is_null"
    IS_NOT_NULL = "is_not_null"
    REGEX = "regex"


class AggregateFunction(str, Enum):
    COUNT = "count"
    SUM = "sum"
    AVG = "avg"
    MIN = "min"
    MAX = "max"
    MEDIAN = "median"
    STD = "std"
    VAR = "var"
    FIRST = "first"
    LAST = "last"
    NUNIQUE = "nunique"


class JoinType(str, Enum):
    INNER = "inner"
    LEFT = "left"
    RIGHT = "right"
    OUTER = "outer"
    CROSS = "cross"


class CleaningRule(BaseModel):
    column: str
    rule_type: str = Field(..., description="Type of cleaning rule")
    parameters: Dict[str, Any] = Field(default_factory=dict)


class FilterRule(BaseModel):
    column: str
    operator: FilterOperator
    value: Any
    case_sensitive: bool = True


class AggregationConfig(BaseModel):
    group_by: List[str] = Field(default_factory=list)
    aggregations: Dict[str, str] = Field(..., description="Column to function mapping")


class JoinConfig(BaseModel):
    right_table: str
    join_type: JoinType = JoinType.INNER
    left_on: List[str]
    right_on: List[str]
    suffixes: Optional[List[str]] = ["_left", "_right"]


class ColumnSplitConfig(BaseModel):
    column: str
    delimiter: str
    new_columns: List[str]
    max_splits: Optional[int] = None


class ColumnMergeConfig(BaseModel):
    columns: List[str]
    new_column: str
    separator: str = " "


class TypeConversionConfig(BaseModel):
    column: str
    target_type: str
    format_string: Optional[str] = None
    errors: str = "coerce"


class CustomScriptConfig(BaseModel):
    script: str
    language: str = "python"
    output_columns: Optional[List[str]] = None


class TransformationStep(BaseModel):
    id: str
    name: str
    type: TransformationType
    enabled: bool = True
    config: Dict[str, Any] = Field(default_factory=dict)
    parameters: Dict[str, Any] = Field(default_factory=dict)
    
    # Type-specific configs (optional, for backward compatibility)
    filter_rules: Optional[List[FilterRule]] = None
    cleaning_rules: Optional[List[CleaningRule]] = None
    aggregation_config: Optional[AggregationConfig] = None
    join_config: Optional[JoinConfig] = None
    split_config: Optional[ColumnSplitConfig] = None
    merge_config: Optional[ColumnMergeConfig] = None
    type_conversion_config: Optional[TypeConversionConfig] = None
    custom_script_config: Optional[CustomScriptConfig] = None


class TransformationPipeline(BaseModel):
    id: Optional[str] = None
    name: str
    description: Optional[str] = None
    source_config: Dict[str, Any]
    steps: List[TransformationStep] = Field(default_factory=list)
    output_config: Optional[Dict[str, Any]] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class TransformationPreviewRequest(BaseModel):
    source_config: Dict[str, Any]
    steps: List[TransformationStep]
    preview_rows: int = 100


class TransformationExecuteRequest(BaseModel):
    pipeline_id: Optional[str] = None
    source_config: Optional[Dict[str, Any]] = None
    steps: Optional[List[TransformationStep]] = None
    output_config: Dict[str, Any]